docs: IBC overview edits and rewrites for clarity (#9001)
cosmos/cosmos-sdk: docs/ibc/overview.md

<!-- order: 1 -->

>  # IBC Overview

Author: a good practice is to use self-describing heads

Learn what IBC is, its components, and use cases. {synopsis}

> ## What is the Inter-Blockchain Communication Protocol (IBC)?

Author: let's use the Inter-Blockchain Communication protocol (IBC) 
capitalize I and B and C and use a hyphen

> This document serves as a guide for developers who want to write their own IBC apps for custom use cases. <!--[use-cases](https://github.com/cosmos/ics/blob/master/ibc/4_IBC_USECASES.md)-->.

Author: this link does not resolve for me, do we have use cases documented somewhere? if not, let's comment out

The modular design of the IBC protocol means that IBC app developers do not require in-depth knowledge of the low-level details of clients, connections, and proof verification. This brief explanation of the lower levels of the stack is provided so that app developers can gain a high-level understanding of the IBC protocol.

The abstraction layer details on channels and ports are relevant for app developers. You can define your own custom packets and IBCModule callbacks.

The following requirements must be met for a module to interact over IBC:

- Bind to one or more ports

- Define the packet data

- Define optional acknowledgement structures and methods to encode and decode them

- Implement the IBCModule interface

## Components Overview

This section describes the IBC components.

> ### [Clients](https://github.com/cosmos/cosmos-sdk/tree/master/x/ibc/core/02-client)

Author: this link is not working for me... I don't see the ibc module on master (is it expected to be on master?) the target file is on https://github.com/cosmos/cosmos-sdk/tree/v0.42.3/x/ibc/core/02-client

### [Proofs](https://github.com/cosmos/cosmos-sdk/tree/master/x/ibc/core/23-commitment) and [Paths](https://github.com/cosmos/cosmos-sdk/tree/master/x/ibc/core/24-host)

In IBC, blockchains do not directly pass messages to each other over the network.

- To communicate, a blockchain commits some state to a precisely defined path reserved for a specific message type and a specific counterparty. For example, a blockchain that stores a specific connectionEnd as part of a handshake or a packet intended to be relayed to a module on the counterparty chain.

- A relayer process monitors for updates to these paths, and relays message that are submitted by the data stored under the path along with a proof to the counterparty chain.

> - The paths that all IBC implementations require for committing IBC messages is defined in [ICS-24 host requirements](https://github.com/cosmos/ics/tree/master/spec/ics-024-host-requirements).

Author: great feedback, thank you!

### [Capabilities](./ocap.md)

IBC is intended to work in execution environments where modules do not necessarily trust each other. IBC must authenticate module actions on ports and channels so that only modules with the appropriate permissions can use the channels. This security is accomplished using [dynamic capabilities](../architecture/adr-003-dynamic-capability-store.md). Upon binding to a port or creating a channel for a module, IBC returns a dynamic capability that the module must claim to use that port or channel. This binding strategy prevents other modules from using that port or channel since those modules do not own the appropriate capability.

While this explanation is useful background information, IBC modules do not need to interact at all with these lower-level abstractions. The relevant abstraction layer for IBC application developers is that of channels and ports.

Write your IBC applications as self-contained **modules**. A module on one blockchain can communicate with other modules on other blockchains by sending, receiving, and acknowledging packets through channels that are uniquely identified by the `(channelID, portID)` tuple.

> A useful analogy is to consider IBC modules as internet apps on a computer. A channel can then be conceptualized as an IP connection, with the IBC portID being like to a IP port, and the IBC channelID being like to an IP address. A single instance of an IBC module can communicate on the same port with any number of other modules and IBC correctly routes all packets to the relevant module using the `(channelID, portID)` tuple. An IBC module can also communicate with another IBC module over multiple ports by sending each `(portID<->portID)` packet stream on a different unique channel.

Author: ```suggestion
A useful analogy is to consider IBC modules as internet apps on a computer. A channel can then be conceptualized as an IP connection, with the IBC portID acting like an IP port, and the IBC channelID acting like an IP address. A single instance of an IBC module can communicate on the same port with any number of other modules and IBC correctly routes all packets to the relevant module using the `(channelID, portID)` tuple. An IBC module can also communicate with another IBC module over multiple ports by sending each `(portID<->portID)` packet stream on a different unique channel.
```

> A useful analogy is to consider IBC modules as internet apps on a computer. A channel can then be conceptualized as an IP connection, with the IBC portID being like to a IP port, and the IBC channelID being like to an IP address. A single instance of an IBC module can communicate on the same port with any number of other modules and IBC correctly routes all packets to the relevant module using the `(channelID, portID)` tuple. An IBC module can also communicate with another IBC module over multiple ports by sending each `(portID<->portID)` packet stream on a different unique channel.

Author: errors on my part, does "acting like" work here?

> A useful analogy is to consider IBC modules as internet apps on a computer. A channel can then be conceptualized as an IP connection, with the IBC portID being like to a IP port, and the IBC channelID being like to an IP address. A single instance of an IBC module can communicate on the same port with any number of other modules and IBC correctly routes all packets to the relevant module using the `(channelID, portID)` tuple. An IBC module can also communicate with another IBC module over multiple ports by sending each `(portID<->portID)` packet stream on a different unique channel.

Reviewer: @barriebyron that sounds good to me :+1:

> A useful analogy is to consider IBC modules as internet apps on a computer. A channel can then be conceptualized as an IP connection, with the IBC portID being like to a IP port, and the IBC channelID being like to an IP address. A single instance of an IBC module can communicate on the same port with any number of other modules and IBC correctly routes all packets to the relevant module using the `(channelID, portID)` tuple. An IBC module can also communicate with another IBC module over multiple ports by sending each `(portID<->portID)` packet stream on a different unique channel.

Reviewer: I saw the "is like" update and that works as well

IBC modules are responsible for claiming the capability that is returned on `BindPort`.

### [Channels](https://github.com/cosmos/cosmos-sdk/tree/master/x/ibc/core/04-channel)

An IBC channel can be established between two IBC ports. A port is exclusively owned by a single module. IBC packets are sent over channels. Just as IP packets contain the destination IP address, IP port, the source IP address, and source IP port, IBC packets contain the destination portID, channelID, the source portID, and channelID. The IBC packets enable IBC to correctly route the packets to the destination module, while also allowing modules receiving packets to know the sender module.

- A channel can be `ORDERED` so that packets from a sending module must be processed by the receiving module in the order they were sent.

- Recommended, a channel may be `UNORDERED` so that packets from a sending module are processed in the order they arrive, which may not be the order the packets were sent.

> Modules may choose which channels they wish to communicate over with. IBC expects modules to implement callbacks that are called during the channel handshake. These callbacks may do custom channel initialization logic, if any return an error, the channel handshake fails. By returning errors on callbacks, modules can programmatically reject and accept channels.

Reviewer: ```suggestion
Modules may choose which channels they wish to communicate over with. IBC expects modules to implement callbacks that are called during the channel handshake. These callbacks may do custom channel initialization logic. If any error is returned, the channel handshake fails. By returning errors on callbacks, modules can programmatically reject and accept channels.
```

The channel handshake is a 4-step handshake. Briefly, if a given chain A wants to open a channel with chain B using an already established connection:

1. Chain A sends a `ChanOpenInit` message to signal a channel initialization attempt with chain B.
2. Chain B sends a `ChanOpenTry` message to try opening the channel on chain A.
3. Chain A sends a `ChanOpenAck` message to mark its channel end status as open.
4. Chain B sends a `ChanOpenConfirm` message to mark its channel end status as open.

> If all of these actions happen successfully, the channel is open on both sides. At each step in the handshake, the module associated with the `ChannelEnd` executes it's callback for that step of the handshake. So on `ChanOpenInit`, the module on chain A has its callback `OnChanOpenInit` executed.

Reviewer: ```suggestion
If all of these actions happen successfully, the channel is open on both sides. At each step in the handshake, the module associated with the `ChannelEnd` executes its callback for that step of the handshake. So on `ChanOpenInit`, the module on chain A has its callback `OnChanOpenInit` executed.
```

In the UNORDERED case, packets can be received in any order. IBC writes a packet receipt for each sequence it has received in the UNORDERED channel. This receipt contains no information and is simply a marker intended to signify that the UNORDERED channel has received a packet at the specified sequence. To timeout a packet on an UNORDERED channel, proof that a packet receipt does not exist is required for the packet's sequence by the specified timeout. Of course, timing out a packet on an UNORDERED channel triggers the application specific timeout logic for that packet, and does not close the channel.

For this reason, most modules that use UNORDERED channels are recommended as they require less liveness guarantees to function effectively for users of that channel.

### [Acknowledgements](https://github.com/cosmos/cosmos-sdk/tree/master/x/ibc/core/04-channel)

Modules also write application-specific acknowledgements when processing a packet. Acknowledgements can be done:

- Synchronously on `OnRecvPacket` if the module processes packets as soon as they are received from IBC module.

- Asynchronously if module processes packets at some later point after receiving the packet.

> This acknowledgement data is opaque to IBC much like the packet `Data` and is treated by IBC as a simple byte string `[]byte`. The receiver modules must encode their acknowledgement so that the sender module can decode it correctly and decided through version negotiation during the channel handshake.

Author: your suggestion is perfect, this was rough content to start, and now we're in a better state, thank you

The acknowledgement can encode whether the packet processing succeeded or failed, along with additional information that allows the sender module to take appropriate action.

> After the acknowledgement has been written by the receiving chain, a relayer relays the acknowledgement back to the original sender module which then executes application-specific acknowledgment logic using the contents of the acknowledgement. This acknowledgement can involve rolling back packet-send changes in the case of a failed acknowledgement (refunding senders).

Author: I think we mean "acknowledgement" in

> This acknowledgement can involve rolling back

Beware the unclear antecedent (what we technical writers call "the naked it" and "the naked this"). It's good practice to repeat the subject

> After an acknowledgement is received successfully on the original sender the chain, the IBC module deletes the corresponding packet commitment as it is no longer needed.

Author: if we mean after, let's use "after" instead of "once" see [Google developer documentation style guide](https://developers.google.com/style/word-list#letter-o)
